lib/imagec/layer_cache.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package imagec
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/vic/lib/apiservers/engine/backends/kv"
+	"github.com/vmware/vic/lib/apiservers/portlayer/client"
+	"github.com/vmware/vic/pkg/trace"
+)
+
+// LCache is an in-memory cache to account for existing image layers
+// It is used primarily by imagec when coordinating layer downloads
+// The cache is initially hydrated by way of the image cache at startup
+type LCache struct {
+	m sync.RWMutex
+	// Layers maps from layer id to layer metadata
+	Layers map[string]*ImageWithMeta
+
+	client *client.PortLayer
+}
+
+// LayerNotFoundError is returned when a layer does not exist in the cache
+type LayerNotFoundError struct{}
+
+func (e LayerNotFoundError) Error() string {
+	return "Layer does not exist"
+}
+
+const (
+	layerCacheKey = "layers"
+)
+
+var (
+	layerCache *LCache
+)
+
+func init() {
+	layerCache = &LCache{
+		Layers: make(map[string]*ImageWithMeta),
+	}
+}
+
+// LayerCache returns a reference to the layer cache
+func LayerCache() *LCache {
+	return layerCache
+}
+
+// NewLayerCache will create a new layer cache or rehydrate an
+// existing layer cache from the portlayer k/v store
+func NewLayerCache(client *client.PortLayer) error {
+	defer trace.End(trace.Begin(""))
+
+	layerCache.client = client
+
+	log.Debugf("Updating layer cache")
+
+	val, err := kv.Get(client, layerCacheKey)
+	if err != nil && err != kv.ErrKeyNotFound {
+		return err
+	}
+
+	if val != "" {
+		if err = json.Unmarshal([]byte(val), layerCache); err != nil {
+			return fmt.Errorf("Failed to unmarshal layer cache: %s", err)
```

Comment:
@hickeng There is currently no solution in place to recover from datastore corruption or version mismatch. It is, however, possible to do a consistency check for the layer and image caches at startup. I think that could be captured in a different issue.

There's also been talk between myself, @hmahmood, @vburenin, @matthewavery about using a distributed k/v store like etcd for this. Perhaps we should continue that discussion?